ashpd-demo/src/portals/desktop/usb.rs

```diff
@@ -0,0 +1,362 @@
+            obj.init_template();
+        }
+    }
+
+    impl ObjectImpl for DeviceRow {}
+    impl WidgetImpl for DeviceRow {}
+    impl ListBoxRowImpl for DeviceRow {}
+    impl PreferencesRowImpl for DeviceRow {}
+    impl ActionRowImpl for DeviceRow {}
+
+    #[derive(Debug, gtk::CompositeTemplate, Default)]
+    #[template(resource = "/com/belmoussaoui/ashpd/demo/usb.ui")]
+    pub struct UsbPage {
+        #[template_child]
+        pub usb_devices: TemplateChild<adw::PreferencesGroup>,
+        rows: RefCell<HashMap<String, super::DeviceRow>>,
+        pub session: Arc<Mutex<Option<Session<'static, UsbProxy<'static>>>>>,
+        pub event_source: Arc<Mutex<Option<glib::SourceId>>>,
+    }
+
+    impl UsbPage {
+        fn add(&self, uuid: String, row: super::DeviceRow) {
+            self.usb_devices.get().add(&row);
+            self.rows.borrow_mut().insert(uuid, row);
+        }
+
+        fn clear_devices(&self) {
+            for row in self.rows.borrow().values() {
+                self.usb_devices.get().remove(row);
+            }
+            self.rows.borrow_mut().clear();
+        }
+
+        fn acquired_device(&self, uuid: &str) {
+            if let Some(row) = self.rows.borrow().get(uuid) {
+                row.acquire();
+            }
+        }
+
+        fn released_device(&self, uuid: &str) {
+            if let Some(row) = self.rows.borrow().get(uuid) {
+                row.release();
+            }
+        }
+
+        pub(super) async fn refresh_devices(&self) -> ashpd::Result<()> {
+            let page = self.obj();
+
+            self.clear_devices();
+
+            let usb = UsbProxy::new().await?;
+            let devices = usb.enumerate_devices(UsbOptions::default()).await?;
+            for device in devices {
+                let row = super::DeviceRow::new();
+                let vendor = device.1.vendor().unwrap_or_default();
+                let dev = device.1.model().unwrap_or_default();
+                row.set_title(&format!("{} {}", &vendor, &dev));
+                if let Some(devnode) = device.1.device_file {
+                    row.set_subtitle(&devnode);
+                }
+
+                let device_id = device.0.clone();
+                let device_writable = device.1.writable.unwrap_or(false);
+                row.connect_share_clicked(clone!(
+                    #[strong]
+                    page,
+                    move |row| {
+                        glib::spawn_future_local(clone!(
+                            #[strong]
+                            row,
+                            #[strong]
+                            device_id,
+                            #[strong]
+                            page,
+                            async move {
+                                let root = row.native().unwrap();
+                                let identifier = WindowIdentifier::from_native(&root).await;
+                                let usb = UsbProxy::new().await.unwrap();
```

Comment:
The unwraps here are not good, once you move it to a separate function, you can make it report errors using the notifications support in PortalPage